File: server/routes/ocr.js
import { Router } from 'express'
import multer from 'multer'
import { promises as fs } from 'fs'
import { existsSync, mkdirSync, writeFileSync } from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { randomUUID } from 'crypto'

const router = Router()

const PADDLEOCR_JOB_URL = 'https://paddleocr.aistudio-app.com/api/v2/ocr/jobs'

function getToken() { return process.env.PADDLEOCR_TOKEN }
function getModel() { return process.env.PADDLEOCR_MODEL || 'PaddleOCR-VL-1.6' }

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const UPLOAD_DIR = path.resolve(__dirname, '..', '..', 'uploads')
const META_FILE = path.join(UPLOAD_DIR, 'photos.json')

// 同步初始化：避免 ESM 顶层 await 兼容问题，并减少启动延迟
if (!existsSync(UPLOAD_DIR)) mkdirSync(UPLOAD_DIR, { recursive: true })
if (!existsSync(META_FILE)) {
  try {
    writeFileSync(META_FILE, '[]', 'utf-8')
  } catch (err) {
    console.warn('[OCR] 初始化 photos.json 失败:', err.message)
  }
}

async function readMeta() {
  try {
    const text = await fs.readFile(META_FILE, 'utf-8')
    return JSON.parse(text || '[]')
  } catch {
    return []
  }
}

async function writeMeta(list) {
  // 先写临时文件再原子 rename，避免写到一半被其他进程读到残缺数据
  const tmpFile = META_FILE + '.tmp'
  await fs.writeFile(tmpFile, JSON.stringify(list, null, 2), 'utf-8')
  await fs.rename(tmpFile, META_FILE)
}

// ---------- 简易 Promise 互斥锁：序列化所有 readMeta→修改→writeMeta 操作 ----------
let _metaLock = Promise.resolve()

/**
 * 在锁保护下执行 meta 读写操作，保证并发安全。
 * @param {function} fn  async (currentList) => newList | void
 *   返回新数组时自动写回；返回 void 表示只读。
 */
function withMetaLock(fn) {
  const next = _metaLock.then(async () => {
    const list = await readMeta()
    const result = await fn(list)
    if (result !== undefined && result !== list) {
      await writeMeta(result)
    }
    return result
  })
  // 不管成败都继续链，防止后续调用被卡死
  _metaLock = next.catch(() => {})
  return next
}

function safeExt(originalName) {
  const ext = path.extname(originalName || '').toLowerCase()
  const allowed = new Set(['.jpg', '.jpeg', '.png', '.bmp', '.webp', '.gif'])
  return allowed.has(ext) ? ext : '.jpg'
}

// 改为磁盘存储：文件落到 uploads/，元数据写到 photos.json
const storage = multer.diskStorage({
  destination: (req, file, cb) => cb(null, UPLOAD_DIR),
  filename: (req, file, cb) => {
    const id = randomUUID()
    const ext = safeExt(file.originalname)
    cb(null, `${id}${ext}`)
  },
})

const upload = multer({
  storage,
  limits: { fileSize: 20 * 1024 * 1024 },
  fileFilter: (req, file, cb) => {
    if (!file.mimetype.startsWith('image/')) {
      return cb(new Error('仅支持图片文件'))
    }
    cb(null, true)
  },
})

/**
 * POST /api/ocr/recognize
 * 接收前端上传的照片：
 *   1) 落盘到 uploads/<uuid>.<ext>
 *   2) 元数据写入 photos.json（status: '识别中'）
 *   3) 提交 PaddleOCR 任务，立即返回 photoId + ocrJobId
 *   4) 后台继续轮询，完成后更新元数据
 */
router.post('/recognize', upload.single('file'), async (req, res) => {
  if (!req.file) {
    return res.status(400).json({ success: false, error: '未上传文件' })
  }

  const photoId = path.basename(req.file.filename, path.extname(req.file.filename))
  const photoUrl = `/uploads/${req.file.filename}`
  const uploadedAt = new Date().toISOString()

  const record = {
    id: photoId,
    originalName: req.file.originalname,
    filename: req.file.filename,
    photoUrl,
    size: req.file.size,
    mimetype: req.file.mimetype,
    uploadedAt,
    ocrStatus: 'pending',
    ocrRawText: '',
    pages: [],
    matchedRuleId: null,
    testItem: '',
    subItem: '',
    recognizedValue: '',
    standardRequirement: '',
    judgment: '待判定',
    status: '识别中',
    includeInReport: false,
    error: null,
    sampleName: '',
    entrustNo: '',
    ocrJobId: null,
  }

  // 先把元数据落库（即使 OCR 失败也保留照片归档）
  await withMetaLock(async (meta) => {
    meta.unshift(record)
    return meta
  })

  try {
    const fileBytes = await fs.readFile(req.file.path)

    const formData = new FormData()
    const blob = new Blob([fileBytes], { type: req.file.mimetype })
    formData.append('file', blob, req.file.originalname)
    formData.append('model', getModel())
    formData.append('optionalPayload', JSON.stringify({
      useDocOrientationClassify: false,
      useDocUnwarping: false,
      useChartRecognition: false,
    }))

    const jobResponse = await fetch(PADDLEOCR_JOB_URL, {
      method: 'POST',
      headers: { 'Authorization': `bearer ${getToken()}` },
      body: formData,
    })

    if (!jobResponse.ok) {
      const text = await jobResponse.text()
      await markFailed(photoId, `PaddleOCR 任务提交失败: ${jobResponse.status}`)
      return res.status(502).json({ success: false, photoId, photoUrl, error: `PaddleOCR 任务提交失败: ${jobResponse.status} ${text}` })
    }

    const jobData = await jobResponse.json()
    const ocrJobId = jobData.data?.jobId

    if (!ocrJobId) {
      await markFailed(photoId, '未获取到 jobId')
      return res.status(502).json({ success: false, photoId, photoUrl, error: '未获取到 jobId' })
    }

    console.log(`[OCR] 任务已提交: ${ocrJobId}, 文件: ${req.file.originalname} → ${req.file.filename}`)

    // 保存 jobId 到元数据
    await withMetaLock(async (list) => {
      const idx = list.findIndex(p => p.id === photoId)
      if (idx >= 0) { list[idx].ocrJobId = ocrJobId; return list }
    })

    // 后台异步轮询，不阻塞当前请求
    pollOCRResult(photoId, ocrJobId, req.file.originalname)

    // 立即返回，前端通过 GET /api/ocr/jobs/:photoId 轮询结果
    return res.json({ success: true, photoId, photoUrl, ocrJobId })
  } catch (err) {
    console.error('[OCR] 服务错误:', err)
    await markFailed(photoId, err.message || '服务内部错误')
    return res.status(500).json({ success: false, photoId, photoUrl, error: err.message || '服务内部错误' })
  }
})

/**
 * GET /api/ocr/jobs/:photoId
 * 前端轮询接口：返回当前照片的 OCR 状态与结果
 */
router.get('/jobs/:photoId', async (req, res) => {
  try {
    const list = await readMeta()
    const item = list.find(p => p.id === req.params.photoId)
    if (!item) return res.status(404).json({ success: false, error: '任务不存在' })

    return res.json({
      success: true,
      photoId: item.id,
      photoUrl: item.photoUrl,
      ocrStatus: item.ocrStatus,
      status: item.status,
      pages: item.pages || [],
      rawText: item.ocrRawText || '',
      tables: item.tables || [],
      error: item.error,
    })
  } catch (err) {
    res.status(500).json({ success: false, error: err.message })
  }
})

/**
 * 后台轮询 PaddleOCR 任务结果（不阻塞 HTTP 请求）
 */
async function pollOCRResult(photoId, ocrJobId, originalName) {
  const MAX_POLLS = 60
  const POLL_INTERVAL = 3000

  try {
    for (let i = 0; i < MAX_POLLS; i++) {
      await new Promise(resolve => setTimeout(resolve, POLL_INTERVAL))

      const pollResponse = await fetch(`${PADDLEOCR_JOB_URL}/${ocrJobId}`, {
        headers: { 'Authorization': `bearer ${getToken()}` },
      })

      if (!pollResponse.ok) continue

      const pollData = await pollResponse.json()
      const state = pollData.data?.state

      if (state === 'done') {
        const jsonUrl = pollData.data?.resultUrl?.jsonUrl
        if (!jsonUrl) {
          await markFailed(photoId, '未获取到结果 URL')
          return
        }

        const resultResponse = await fetch(jsonUrl)
        if (!resultResponse.ok) {
          await markFailed(photoId, '获取识别结果失败')
          return
        }

        const resultText = await resultResponse.text()
        const parsed = parseOCRResult(resultText)

        await markDone(photoId, parsed)
        console.log(`[OCR] 识别完成: ${originalName}, 文本长度: ${parsed.rawText.length}, 表格数: ${(parsed.tables || []).length}`)
        return
      }

      if (state === 'failed') {
        const errorMsg = pollData.data?.errorMsg || '识别失败'
        await markFailed(photoId, errorMsg)
        return
      }
    }

    await markFailed(photoId, '识别超时')
  } catch (err) {
    console.error(`[OCR] 后台轮询异常 (${photoId}):`, err)
    await markFailed(photoId, err.message || '后台轮询异常')
  }
}

async function markFailed(photoId, errorMsg) {
  await withMetaLock(async (list) => {
    const idx = list.findIndex(p => p.id === photoId)
    if (idx >= 0) {
      list[idx] = {
        ...list[idx],
        ocrStatus: 'failed',
        status: '识别失败',
        error: errorMsg,
      }
      return list
    }
  })
}

async function markDone(photoId, parsed) {
  await withMetaLock(async (list) => {
    const idx = list.findIndex(p => p.id === photoId)
    if (idx >= 0) {
      list[idx] = {
        ...list[idx],
        ocrStatus: 'done',
        status: '待确认',
        pages: parsed.pages,
        ocrRawText: parsed.rawText,
        tables: parsed.tables || [],
        error: null,
      }
      return list
    }
  })
}

function parseOCRResult(jsonlText) {
  const lines = jsonlText.trim().split('\n').filter(Boolean)
  const pages = []
  let rawText = ''
  // 合并所有页 markdown 之后再解析表格，避免 PaddleOCR 把一个 <tr> 拆到多段
  let combinedMarkdown = ''

  for (const line of lines) {
    try {
      const parsed = JSON.parse(line)
      const results = parsed.result?.layoutParsingResults || []
      for (const res of results) {
        const pageText = res.markdown?.text || ''
        rawText += pageText + '\n'
        combinedMarkdown += pageText + '\n'
        pages.push({
          pageNumber: pages.length + 1,
          text: pageText,
          images: res.markdown?.images || {},
        })
      }
    } catch {
      // skip malformed lines
    }
  }

  // 解析 markdown 里的 HTML 表格，产出结构化表格数据
  const tables = parseHtmlTables(combinedMarkdown)

  return {
    pages,
    rawText: rawText.trim(),
    tables, // [{ rows: [[cell, cell, ...], ...], headers?: [string] }]
  }
}

/**
 * 从 markdown 文本中解析 <table>...</table>。
 * 兼容 PaddleOCR 偶发的截断/拼接异常：
 *   - 同一行可能被切到两段 markdown 字符串里（>200.00</td> 这种残尾）
 *   - 单元格里有 <br> / &nbsp; / 前后空白
 * 返回：[{ rows: [[cell, cell, ...], ...] }]
 */
function parseHtmlTables(markdown) {
  const tables = []
  if (!markdown || typeof markdown !== 'string') return tables

  // 先把跨段被切断的 <td> 残尾粘回去：把 ">\d+(\.\d+)?</td>" 形式
  // 重新规范为 "<td>数字</td>"，避免解析器把同一格当成两段
  const normalized = markdown
    .replace(/<br\s*\/?>/gi, ' ')
    .replace(/&nbsp;/gi, ' ')
    .replace(/>\s*(\d+(?:\.\d+)?)\s*<\/td>/g, '<td>$1</td>')
    .replace(/>\s*<\/td>/g, '></td>')

  const tableRegex = /<table[^>]*>([\s\S]*?)<\/table>/gi
  let match
  while ((match = tableRegex.exec(normalized)) !== null) {
    const tableHtml = match[1]
    const rowRegex = /<tr[^>]*>([\s\S]*?)<\/tr>/gi
    const rows = []
    let rowMatch
    while ((rowMatch = rowRegex.exec(tableHtml)) !== null) {
      const rowHtml = rowMatch[1]
      // 收集 <td> 和 <th>，保留 colspan
      const cellRegex = /<(td|th)\b[^>]*colspan\s*=\s*["']?(\d+)[^>]*>([\s\S]*?)<\/\1>|<(td|th)\b[^>]*>([\s\S]*?)<\/\4>/gi
      const cells = []
      let cellMatch
      while ((cellMatch = cellRegex.exec(rowHtml)) !== null) {
        const span = cellMatch[2] ? parseInt(cellMatch[2], 10) : 1
        const raw = (cellMatch[3] ?? cellMatch[5] ?? '').toString()
        const text = stripHtml(raw).trim()
        for (let i = 0; i < span; i++) cells.push(text)
      }
      if (cells.length > 0) rows.push(cells)
    }
    if (rows.length > 0) tables.push({ rows })
  }
  return tables
}

function stripHtml(s) {
  return s
    .replace(/<[^>]+>/g, '')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
}

/**
 * GET /api/ocr/photos
 * 查询归档照片列表。可选 query：
 *   sampleName, entrustNo, includeInReport (true/false), status
 */
router.get('/photos', async (req, res) => {
  try {
    const list = await readMeta()
    const { sampleName, entrustNo, includeInReport, status } = req.query

    let filtered = list
    if (sampleName) filtered = filtered.filter(p => p.sampleName === sampleName)
    if (entrustNo) filtered = filtered.filter(p => p.entrustNo === entrustNo)
    if (includeInReport === 'true') filtered = filtered.filter(p => p.includeInReport)
    if (status) filtered = filtered.filter(p => p.status === status)

    res.json({ success: true, total: filtered.length, items: filtered })
  } catch (err) {
    res.status(500).json({ success: false, error: err.message })
  }
})

/**
 * GET /api/ocr/photos/:id
 */
router.get('/photos/:id', async (req, res) => {
  const list = await readMeta()
  const item = list.find(p => p.id === req.params.id)
  if (!item) return res.status(404).json({ success: false, error: '照片不存在' })
  res.json({ success: true, item })
})

/**
 * PATCH /api/ocr/photos/:id
 * 更新元数据：testItem / subItem / recognizedValue / judgment / status /
 *   includeInReport / sampleName / entrustNo / matchedRuleId
 */
router.patch('/photos/:id', async (req, res) => {
  try {
    const result = await withMetaLock(async (list) => {
      const idx = list.findIndex(p => p.id === req.params.id)
      if (idx < 0) return null

      const editable = [
        'testItem', 'subItem', 'recognizedValue', 'standardRequirement',
        'judgment', 'status', 'includeInReport', 'sampleName', 'entrustNo',
        'matchedRuleId', 'matchedRuleName',
      ]
      const patch = {}
      for (const k of editable) {
        if (k in req.body) patch[k] = req.body[k]
      }
      patch.updatedAt = new Date().toISOString()
      list[idx] = { ...list[idx], ...patch }
      return list
    })

    if (result === null) {
      return res.status(404).json({ success: false, error: '照片不存在' })
    }
    const item = result.find(p => p.id === req.params.id)
    res.json({ success: true, item })
  } catch (err) {
    res.status(500).json({ success: false, error: err.message })
  }
})

/**
 * DELETE /api/ocr/photos/:id
 * 删除归档（磁盘 + 元数据）
 */
router.delete('/photos/:id', async (req, res) => {
  try {
    let filePath = null
    const result = await withMetaLock(async (list) => {
      const idx = list.findIndex(p => p.id === req.params.id)
      if (idx < 0) return null
      const item = list[idx]
      filePath = path.join(UPLOAD_DIR, item.filename)
      list.splice(idx, 1)
      return list
    })

    if (result === null) {
      return res.status(404).json({ success: false, error: '照片不存在' })
    }

    if (filePath) {
      try {
        await fs.unlink(filePath)
      } catch (err) {
        console.warn(`[OCR] 删除文件失败: ${filePath}`, err.message)
      }
    }

    res.json({ success: true })
  } catch (err) {
    res.status(500).json({ success: false, error: err.message })
  }
})

/**
 * GET /uploads/:filename  - 由 server/index.js 在路由层挂载，此处不重复挂载
 * 但保留一个下载端点以便前端触发下载（设置 Content-Disposition）
 */
router.get('/photos/:id/download', async (req, res) => {
  const list = await readMeta()
  const item = list.find(p => p.id === req.params.id)
  if (!item) return res.status(404).json({ success: false, error: '照片不存在' })

  const filePath = path.join(UPLOAD_DIR, item.filename)
  res.download(filePath, item.originalName)
})

export default router

File: backend/app/services/ocr.py
import json
import asyncio
import logging
import uuid
import httpx
from datetime import datetime
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.config import settings
from app.services.table_parser import parse_html_tables
from app.services.extractors import get_extractor, get_extractor_for_config
from app.models.photo import Photo

logger = logging.getLogger(__name__)

PADDLEOCR_JOB_URL = "https://paddleocr.aistudio-app.com/api/v2/ocr/jobs"
MAX_POLLS = 100
POLL_INTERVAL = 2.0


def _parse_ocr_result(jsonl_text: str) -> dict:
    lines = [l for l in jsonl_text.strip().split("\n") if l.strip()]
    pages = []
    raw_text = ""

    for line in lines:
        try:
            parsed = json.loads(line)
            results = parsed.get("result", {}).get("layoutParsingResults", [])
            for res in results:
                page_text = res.get("markdown", {}).get("text", "")
                raw_text += page_text + "\n"
                pages.append({
                    "page_number": len(pages) + 1,
                    "text": page_text,
                    "images": res.get("markdown", {}).get("images", {}),
                })
        except (json.JSONDecodeError, KeyError):
            continue

    # Parse HTML tables from combined markdown
    tables = parse_html_tables(raw_text)

    return {"pages": pages, "raw_text": raw_text.strip(), "tables": tables}


def compress_image(file_bytes: bytes, filename: str, content_type: str, max_size: int = 5 * 1024 * 1024) -> tuple[bytes, str, str]:
    """压缩超过 max_size 的图片，返回 (bytes, filename, content_type)。

    使用 Pillow 按比例缩小图片直到小于 max_size，重新编码为 JPEG。
    如果图片本身不大或 Pillow 不可用，返回原始数据。
    """
    if len(file_bytes) <= max_size:
        return file_bytes, filename, content_type

    try:
        from PIL import Image
        import io

        img = Image.open(io.BytesIO(file_bytes))
        # 转为 RGB（去掉 alpha 通道）
        if img.mode in ("RGBA", "P"):
            img = img.convert("RGB")

        # 逐步缩小尺寸
        quality = 85
        output = io.BytesIO()
        img.save(output, format="JPEG", quality=quality)
        while output.tell() > max_size and quality > 30:
            output.seek(0)
            output.truncate()
            quality -= 10
            img.save(output, format="JPEG", quality=quality)

        if output.tell() > max_size:
            # 还是不够小，缩小尺寸
            ratio = (max_size / output.tell()) ** 0.5
            new_size = (int(img.width * ratio), int(img.height * ratio))
            img = img.resize(new_size, Image.LANCZOS)
            output.seek(0)
            output.truncate()
            img.save(output, format="JPEG", quality=60)

        compressed = output.getvalue()
        logger.info(f"图片压缩: {len(file_bytes)} -> {len(compressed)} bytes (quality={quality})")
        new_name = filename.rsplit(".", 1)[0] + ".jpg"
        return compressed, new_name, "image/jpeg"

    except ImportError:
        logger.warning("Pillow 未安装，跳过图片压缩")
        return file_bytes, filename, content_type
    except Exception as e:
        logger.warning(f"图片压缩失败: {e}，使用原始图片")
        return file_bytes, filename, content_type


async def submit_ocr_job(file_bytes: bytes, filename: str, content_type: str) -> str | None:
    """Submit OCR job to PaddleOCR and return the job ID."""
    # 压缩大图片
    file_bytes, filename, content_type = compress_image(file_bytes, filename, content_type)

    async with httpx.AsyncClient(timeout=60) as client:
        files = {"file": (filename, file_bytes, content_type)}
        data = {
            "model": settings.paddleocr_model,
            "optionalPayload": json.dumps({
                "useDocOrientationClassify": False,
                "useDocUnwarping": False,
                "useChartRecognition": False,
            }),
        }
        headers = {"Authorization": f"bearer {settings.paddleocr_token}"}

        resp = await client.post(PADDLEOCR_JOB_URL, headers=headers, files=files, data=data)
        if resp.status_code != 200:
            logger.error(f"PaddleOCR submit failed: {resp.status_code} {resp.text}")
            return None

        job_data = resp.json()
        return job_data.get("data", {}).get("jobId")


async def poll_ocr_and_update_db(photo_id: str, job_id: str, db: AsyncSession):
    """Background task: poll PaddleOCR for result and update DB record."""
    headers = {"Authorization": f"bearer {settings.paddleocr_token}"}

    for i in range(MAX_POLLS):
        await asyncio.sleep(POLL_INTERVAL)

        try:
            async with httpx.AsyncClient(timeout=30) as client:
                poll_resp = await client.get(
                    f"{PADDLEOCR_JOB_URL}/{job_id}",
                    headers=headers,
                )
                if poll_resp.status_code != 200:
                    continue

                poll_data = poll_resp.json()
                state = poll_data.get("data", {}).get("state")

                if state == "done":
                    json_url = poll_data.get("data", {}).get("resultUrl", {}).get("jsonUrl")
                    if not json_url:
                        await _mark_failed(db, photo_id, "未获取到结果 URL")
                        return

                    result_resp = await client.get(json_url)
                    if result_resp.status_code != 200:
                        await _mark_failed(db, photo_id, "获取识别结果失败")
                        return

                    parsed = _parse_ocr_result(result_resp.text)

                    async with db.begin():
                        result = await db.execute(select(Photo).where(Photo.id == photo_id))
                        photo = result.scalar_one_or_none()
                        if photo:
                            photo.ocr_status = "completed"
                            photo.status = "待确认"
                            photo.ocr_raw_text = parsed["raw_text"]
                            photo.pages = parsed["pages"]
                            photo.tables = [t for t in (parsed["tables"] or [])]
                            photo.updated_at = datetime.utcnow()

                    logger.info(f"OCR completed for photo {photo_id}")
                    return

                if state == "failed":
                    error_msg = poll_data.get("data", {}).get("errorMsg", "识别失败")
                    await _mark_failed(db, photo_id, error_msg)
                    return

        except Exception as e:
            logger.error(f"OCR poll error for {photo_id}: {e}")
            continue

    # Timeout
    await _mark_failed(db, photo_id, "识别超时")


async def _mark_failed(db: AsyncSession, photo_id: str, error: str):
    try:
        async with db.begin():
            result = await db.execute(select(Photo).where(Photo.id == photo_id))
            photo = result.scalar_one_or_none()
            if photo:
                photo.ocr_status = "failed"
                photo.status = "识别失败"
                photo.error = error
                photo.updated_at = datetime.utcnow()
        logger.warning(f"OCR failed for photo {photo_id}: {error}")
    except Exception as e:
        logger.error(f"Failed to mark photo {photo_id} as failed: {e}")


async def poll_ocr_and_update_db_with_config(
    photo_id: str,
    job_id: str,
    db: AsyncSession,
    device_key: str,
    sample_name: str,
    group_id: str,
    configs: list[dict],
):
    """带配置的 OCR 轮询：OCR 完成后按配置组批量提取多个检测项。

    Args:
        photo_id: 主照片 ID（OCR 结果存于此）
        job_id: PaddleOCR job ID
        db: 异步数据库会话
        device_key: 设备编号
        sample_name: 样品名称
        group_id: 检测项配置组 ID（本次测试共享）
        configs: 该设备+样品的配置列表（来自 config_cache）
    """
    headers = {"Authorization": f"bearer {settings.paddleocr_token}"}

    for i in range(MAX_POLLS):
        await asyncio.sleep(POLL_INTERVAL)

        try:
            async with httpx.AsyncClient(timeout=30) as client:
                poll_resp = await client.get(
                    f"{PADDLEOCR_JOB_URL}/{job_id}",
                    headers=headers,
                )
                if poll_resp.status_code != 200:
                    continue

                poll_data = poll_resp.json()
                state = poll_data.get("data", {}).get("state")

                if state == "done":
                    json_url = poll_data.get("data", {}).get("resultUrl", {}).get("jsonUrl")
                    if not json_url:
                        await _mark_failed(db, photo_id, "未获取到结果 URL")
                        return

                    result_resp = await client.get(json_url)
                    if result_resp.status_code != 200:
                        await _mark_failed(db, photo_id, "获取识别结果失败")
                        return

                    parsed = _parse_ocr_result(result_resp.text)
                    tables = parsed.get("tables", []) or []
                    raw_text = parsed.get("raw_text", "")

                    # 按配置组批量提取
                    photo_records = []

                    if configs:
                        for cfg in configs:
                            try:
                                # 按配置的 extraction_rule.type 选择提取器（优先），
                                # fallback 到 device_key 映射
                                extractor = get_extractor_for_config(cfg)
                                if not extractor:
                                    logger.warning(
                                        f"No extractor for config {cfg.get('id')} "
                                        f"(type={cfg.get('extraction_rule', {}).get('type', '?')}, "
                                        f"device={device_key}), skip extraction"
                                    )
                                    raise ValueError("no matching extractor")

                                extract_result = extractor.extract(tables, raw_text, cfg)
                                values = extract_result.get("values", [])
                                result_val = extract_result.get("result")
                                frequency_data = extract_result.get("frequency_data")

                                # 提取返回空结果时设置错误提示，触发手动填写
                                extract_error = None
                                if result_val is None and not values:
                                    extract_error = "未提取到数据，请手动填写"

                                # 每个检测项生成一条 Photo 记录，共享 group_id
                                child_photo = Photo(
                                    id=uuid.uuid4().hex,
                                    original_name=f"{cfg['test_item']}",
                                    filename=f"child_{uuid.uuid4().hex}",  # 子记录用唯一标识，避免空字符串冲突
                                    photo_url="",
                                    size=0,
                                    mimetype="",
                                    ocr_status="completed",
                                    status="待确认",
                                    error=extract_error,
                                    ocr_raw_text=raw_text if cfg == configs[0] else None,
                                    pages=parsed.get("pages") if cfg == configs[0] else None,
                                    tables=tables if cfg == configs[0] else None,
                                    test_item=cfg["test_item"],
                                    sub_item=cfg.get("sub_item"),
                                    recognized_value=str(result_val) if result_val is not None else None,
                                    standard_requirement=cfg.get("judgment_indicator"),
                                    judgment="待判定",
                                    sample_name=sample_name,
                                    device_key=device_key,
                                    group_id=group_id,
                                    config_id=cfg["id"],
                                    material_spec=cfg.get("material_spec"),
                                    sample_count=cfg.get("sample_count"),
                                    aggregation_method=cfg.get("aggregation_method"),
                                    result_values=values,
                                    frequency_data=frequency_data,
                                )
                                photo_records.append(child_photo)
                            except Exception as e:
                                logger.warning(
                                    f"Extractor {device_key} failed for config {cfg.get('id')}: {e}"
                                )
                                # 失败的检测项仍创建记录，标记需手动填写
                                child_photo = Photo(
                                    id=uuid.uuid4().hex,
                                    original_name=f"{cfg['test_item']}",
                                    filename=f"child_{uuid.uuid4().hex}",  # 子记录用唯一标识
                                    photo_url="",
                                    size=0,
                                    mimetype="",
                                    ocr_status="completed",
                                    status="待确认",
                                    test_item=cfg["test_item"],
                                    sub_item=cfg.get("sub_item"),
                                    recognized_value=None,
                                    standard_requirement=cfg.get("judgment_indicator"),
                                    judgment="待判定",
                                    error="未提取到数据，请手动填写",
                                    sample_name=sample_name,
                                    device_key=device_key,
                                    group_id=group_id,
                                    config_id=cfg["id"],
                                    material_spec=cfg.get("material_spec"),
                                    sample_count=cfg.get("sample_count"),
                                    aggregation_method=cfg.get("aggregation_method"),
                                    result_values=[],
                                )
                                photo_records.append(child_photo)
                    else:
                        # 无配置：回退为通用模式，仅存原始数据
                        logger.warning(
                            f"No configs for device {device_key} + sample {sample_name}, fallback to raw mode"
                        )

                    # 更新主照片记录（保留 OCR 原始结果）
                    async with db.begin():
                        result = await db.execute(select(Photo).where(Photo.id == photo_id))
                        photo = result.scalar_one_or_none()
                        if photo:
                            photo.ocr_status = "completed"
                            photo.status = "待确认"
                            photo.ocr_raw_text = raw_text
                            photo.pages = parsed.get("pages")
                            photo.tables = tables
                            photo.device_key = device_key
                            photo.group_id = group_id
                            photo.sample_name = sample_name
                            photo.updated_at = datetime.utcnow()

                    # 批量插入子记录
                    if photo_records:
                        async with db.begin():
                            for rec in photo_records:
                                db.add(rec)

                    logger.info(
                        f"OCR with config completed for photo {photo_id}, "
                        f"extracted {len(photo_records)} items"
                    )
                    return

                if state == "failed":
                    error_msg = poll_data.get("data", {}).get("errorMsg", "识别失败")
                    await _mark_failed(db, photo_id, error_msg)
                    return

        except Exception as e:
            logger.error(f"OCR poll error for {photo_id}: {e}")
            continue

    # Timeout
    await _mark_failed(db, photo_id, "识别超时")
